File: apps/api/src/models/HumanTask.ts
import {
  Model,
  DataTypes,
  Optional,
  BelongsToGetAssociationMixin,
} from "sequelize";
import type { ModelQuery } from "./ModelQuery";
import { sequelize } from "../config/database";

export type HumanTaskStatus = "PENDING" | "RESOLVED";

export interface HumanTaskAttributes {
  id: string;
  modelQueryId: string;
  reason: string;
  retrievedData?: string | null;
  modelMessage: string;
  status: HumanTaskStatus;
  createdAt?: Date;
  updatedAt?: Date;
  isDeleted?: boolean;
  deletedAt?: Date | null;
}

export type HumanTaskCreationAttributes = Optional<
  HumanTaskAttributes,
  | "id"
  | "status"
  | "retrievedData"
  | "createdAt"
  | "updatedAt"
  | "isDeleted"
  | "deletedAt"
>;

export class HumanTask
  extends Model<HumanTaskAttributes, HumanTaskCreationAttributes>
  implements HumanTaskAttributes
{
  declare id: string;
  declare modelQueryId: string;
  declare reason: string;
  declare retrievedData: string | null;
  declare modelMessage: string;
  declare status: HumanTaskStatus;
  declare readonly createdAt: Date;
  declare readonly updatedAt: Date;
  declare isDeleted: boolean;
  declare deletedAt: Date | null;

  declare getModelQuery: BelongsToGetAssociationMixin<ModelQuery>;
}

HumanTask.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    modelQueryId: {
      type: DataTypes.UUID,
      allowNull: false,
      unique: true,
      field: "model_query_id",
    },
    reason: {
      type: DataTypes.TEXT,
      allowNull: false,
    },
    retrievedData: {
      type: DataTypes.TEXT,
      allowNull: true,
      field: "retrieved_data",
    },
    modelMessage: {
      type: DataTypes.TEXT,
      allowNull: false,
      field: "model_message",
    },
    status: {
      type: DataTypes.ENUM("PENDING", "RESOLVED"),
      allowNull: false,
      defaultValue: "PENDING",
    },
    isDeleted: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      field: "is_deleted",
    },
    deletedAt: {
      type: DataTypes.DATE,
      allowNull: true,
      field: "deleted_at",
    },
  },
  {
    sequelize,
    tableName: "human_tasks",
  },
);

File: python/app/models/human_task.py
"""HumanTask model: tasks for human review linked to a model query."""

from __future__ import annotations

import uuid

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.models.base import Base


class HumanTask(Base):
    __tablename__ = "human_tasks"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    model_query_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("model_queries.id", onupdate="CASCADE", ondelete="CASCADE"),
        nullable=False,
        unique=True,
        index=True,
    )
    reason: Mapped[str] = mapped_column(Text(), nullable=False)
    retrieved_data: Mapped[str | None] = mapped_column(Text(), nullable=True)
    model_message: Mapped[str] = mapped_column(Text(), nullable=False)
    status: Mapped[str] = mapped_column(String(), nullable=False, server_default="PENDING")  # PENDING, RESOLVED
    created_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )
    is_deleted: Mapped[bool] = mapped_column(Boolean(), server_default="false", nullable=False)
    deleted_at: Mapped[DateTime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    model_query: Mapped["ModelQuery"] = relationship("ModelQuery", back_populates="human_task")  # noqa: F821
